### ISSUE TITLE
Let's get our best results in Graphalytics

### ISSUE THREAD

**Issue by jowens** [gunrock/gunrock#651]
https://graphalytics.org/

Not sure how good their results are, but they don't seem complete for Gunrock.

Good new-student project.

**Comment by bashdash4** [comment]
Is the Platform Driver for Gunrock available? I can't get it from the Graphalytics website as it's listed as "Currently private" (https://graphalytics.org/software).

**Comment by jowens** [comment]
I don't even know what a "platform driver" is. We did not write one.

**Comment by bashdash4** [comment]
There is one dataset that is in both our repo and their Datasets page: cit-Patents, which features 3774768 vertices and 16518947 edges.

**Comment by jowens** [comment]
OK. For posterity, what are their other datasets? Helpful if you find a link for them on, say, snap.stanford.edu or the SuiteSparse pages.

**Comment by jowens** [comment]
What are you stuck on re: the platform driver? I would bet it’s glue code that they wrote that bridges Gunrock and whatever common code they have for evaluation.

**Comment by bashdash4** [comment]
Of course, here is the link to their dataset page (https://graphalytics.org/datasets).

I will let you know if I come across any obstacles with the development of the driver.

**Comment by neoblizz** [comment]
Is there an example of an existing driver for another framework? Maybe we already have it.

**Comment by bashdash4** [comment]
The existing drivers can be found on the software page (https://graphalytics.org/software).

**Comment by bashdash4** [comment]
Regarding contacting the Graphalytics team regarding graph format information:

The site doesn't have contact information (I could attempt to communicate through GitHub in case we need to make sure we have the right idea). However, from the "Graph Loading and Unloading" section in the guide for creating a platform driver (https://github.com/ldbc/ldbc_graphalytics/wiki/Manual%3A-Implementing-Driver#graph-loading-and-unloading), it says that the platform should be responsible for pre-processing, i.e. converting graph data to a suitable format, if it can't use the graph data as is.

Though it's definitely possible to set up the graph loading function of the driver so that the driver handles the conversion through an external process, I believe the expectation is that Gunrock itself will handle .e and .v files, either by using them directly or converting them to Matrix Market file format...

**Comment by bashdash4** [comment]
https://github.com/ldbc/ldbc_graphalytics/issues/194

**Comment by neoblizz** [comment]
I had the pleasure of meeting @szarnyasg at FOSDEM. He mentioned we can contact him to obtain matrix market format files, he has all of them converted already for the required Graphalytics benchmarking.

**Comment by neoblizz** [comment]
> Hi Muhammad,
> 
> thanks for picking this up so quickly. Let's get this rolling!
> 
> I attached two files containing the links to my Dropbox. As you have probably noticed, some graphs in Graphalytics have edge weights and some don't. The '...-default.txt' file has matrices for all graphs - the graphs with edge weights are stored as "real" matrices with the edge weights encoded. Loading edge weights is unnecessary for some algorithms, in these cases, I recommend the matrices in the '-bool-only.txt' file.
> 
> The files are compressed with ZSTD. I usually decompress them on the fly using:
> 
> while read ZSTD_FILE_URL; do
>     curl -L $ZSTD_FILE_URL | tar -I zstd -xv
> done < dropbox-matrix-market-default.txt
> 
> I'll also chime in the issue. Let me know if you have any questions regarding implementing a platform driver.
> I think your best bet would be to start off with the GraphBLAS driver: https://github.com/ftsrg/ldbc-graphalytics-platform-graphblas. This is well maintained and has some similarity to Gunrock (unlike, say, the distributed Java-based implementations).
> 
> Best regards,
> Gabor


> I forgot to mention that Dropbox isn't blazing fast and it does a bit of throttling when downloading large files. But overall, it should be fine for this size. The files are 200G compressed and ~1.5TB decompressed.
> Here's a summary of the data used: https://docs.google.com/spreadsheets/d/e/2PACX-1vTuNlJDe521a21Conz5IjPiDwVvd0Rrd_WwJ7aUkszABDi9lvIwezesvzn2S2XblELa7wUAtisbu-jF/pubhtml

> We managed to dig out the Gunrock platform driver. It was in a private repo of someone who has since left the project :-).
> I'm attaching the source code. Don't expect too much from it though - this was a very preliminary piece of code (I thin that's why it was not published), so I think your best bet is still to start with the Graphalytics driver.
> (https://github.com/gunrock/gunrock/files/4186223/graphalytics-platforms-gunrock-master.1.zip)
> 

[graphalytics-platforms-gunrock-master.zip](https://github.com/gunrock/gunrock/files/4186227/graphalytics-platforms-gunrock-master.1.zip)

Gabor was kind enough to provide us with all this information and the datasets. He asked me to keep the dropbox links private, so I will go ahead and download the datasets myself on the servers for us to use.

**Comment by achalagarwal** [comment]
@szarnyasg 
Hi Gabor,

I was wondering if there is a specific reason why graphalytics has its own timer for load graph
[load-graph_timer](https://github.com/ldbc/ldbc_graphalytics/blob/094f5a54039ea47cefb3316824f0ea13e1ceecc8/graphalytics-core/src/main/java/science/atlarge/graphalytics/execution/BenchmarkExecutor.java#L135) but uses platform supplied metrics for the algorithm run time. 

For us (gunrock), it would be easier to benchmark our results if we could supply load-graph metrics separately as our APIs are not disintegrated as such (or we can perform a load graph twice, once for the load-graph metrics and the second time for running the algorithm but subtract this time to reflect the actual algorithm running time).

Cheers

**Comment by szarnyasg** [comment]
Hello @achalagarwal,

apologies, I missed the notification for this question.

The JSON format in Graphalytics is legacy stuff, and I'm personally not a big fan of it. In fact, I immediately convert it to CSVs with the following R script:

```r
read_gx_json = function(filename, conf) {
  head = c("platform", "algorithm", "dataset", "job", "run", "load", "make", "proc", "conf")
  df = data.frame()
  j = read_json(filename)
  jobs = j[["result"]][["jobs"]]
  for (job in jobs) {
    job_id = job[["id"]]
    algorithm = job[["algorithm"]]
    dataset = job[["dataset"]]
    run_ids = job[["runs"]]
    
    for (run_id in run_ids) {
      run = j[["result"]][["runs"]][[run_id]]
      id = run[["id"]]
      load = as.double(run[["load_time"]])
      make = as.double(run[["makespan"]])
      proc = as.double(run[["processing_time"]])
      
      row = data.frame(j[["system"]][["platform"]][["name"]], algorithm, dataset, job_id, run_id, load, make, proc, conf)
      names(row) = head
      df = rbind(df, row)
    }
  }
  df
}

# throw away job/run ids
df = df[ , !(names(df) %in% c("job", "run"))]

# compute load time and drop makespan time
df$load = df$make - df$proc
df = df[ , !(names(df) %in% c("make"))]

# aggregate and melt for visualization
ag = df
ag = ddply(.data = ag, .variables = c("platform", "algorithm", "dataset", "conf"), .fun = colwise(median))
md <- melt(ag, id.vars=c("platform", "algorithm", "dataset", "conf"))

df = rbind(
  read_gx_json("results-grb310.json",  "GraphBLAS-v3.1.0"),
  read_gx_json("results-grb320.json",  "GraphBLAS-v3.2.0")
)
```

The specifics of this code are not super important - the bottom line is that if you can provide CSVs with schema `"platform", "algorithm", "dataset", "load", "proc", "conf"` - or something similar with `make` (makespan) instead of `load` -, I can use, visualize, and analyze that.


Gabor

**Comment by achalagarwal** [comment]
Thank you @szarnyasg for the script and the clarification.

From what I gather then, I can ignore the load attribute of the generated report when creating the CSV.

Also,  I had been going through at Graphalytics 1.2.0. Is it alright if I continue with that or is it critical to move ahead with 1.3.0?

**Comment by szarnyasg** [comment]
@achalagarwal Graphalytics v1.3.0 has minimal changes: one fixed configuration file and improved logging in case of validation errors: https://github.com/ldbc/ldbc_graphalytics/releases

Making the switch should be fairly easy but not necessary so it's up to you.

**Comment by szarnyasg** [comment]
@achalagarwal how's your progress going with Gunrock × Graphalytics? Can you give an estimate which algorithms you'll be able to support - all six directed/undirected - and when an initial implementation will be ready?

Thanks,
Gabor

**Comment by achalagarwal** [comment]
Hi @szarnyasg, the connection between gunrock and graphalytics is done. Which means that the experimentation remains (which is going to take time, I do not have a good estimate for you as it involves parameter-tuning). 

Only the connected components algorithm would take some time but that depends on when I can finish the experiments for the remaining five algorithms on the other graphs. We work with directed graphs, and undirected graphs are converted to directed graphs (which is a preprocessing step and would thus go in loadtime I guess)

Are you okay with incremental updates to the benchmark results using custom benchmarks? To be specific, can I send you results for a subset of algorithms + subset of datasets initially and then send the ones missing?

**Comment by szarnyasg** [comment]
@achalagarwal Great. Is the code available somewhere?

> (which is a preprocessing step and would thus go in loadtime I guess)
Yes.

> Are you okay with incremental updates to the benchmark results using custom benchmarks? To be specific, can I send you results for a subset of algorithms + subset of datasets initially and then send the ones missing?
Sure! I can stitch together partial results.

**Comment by achalagarwal** [comment]
@szarnyasg currently I was working off the graphblas skeleton and adding/modifying code as required. I do plan to setup a clean driver repository for gunrock but didn't prioritize it. 

I guess I should then so that you could have a look if needed! Will try to do it this weekend.

**Comment by szarnyasg** [comment]
@achalagarwal - it's not super urgent, feel free to take your time.

**Comment by achalagarwal** [comment]
@szarnyasg Hi, I generated a sample results file, please let me know if the csv/json (files attached) format works for you. 

[sample-results.tar.gz](https://github.com/gunrock/gunrock/files/4364358/sample-results.tar.gz)

**Comment by szarnyasg** [comment]
@achalagarwal thanks! I'm using the CSV and it's almost confirming my script:

- I don't need the first column (no big deal, I'm just throwing it away)
- the `load` column in missing

**Comment by achalagarwal** [comment]
> provide CSVs with schema "platform", "algorithm", "dataset", "load", "proc", "conf" - or something similar with make (makespan) instead of load

@szarnyasg 
did you mean that I use the makespan value in the "load" column?

Also do I assume that for algorithms where disconnected graphs don't make too much sense (like bfs) the benchmark graph datasets will be connected?

**Comment by szarnyasg** [comment]
> did you mean that I provide the makespan value in the "load" column?

If possible, you should put the preprocessing time into `load` column. According to the spec:

"During the Loading step, the input graph data can be converted into an optimized system-specific data format and pre-loaded into a local/share/distributed storage system. Binary formats are allowed, however, the preprocessed data must be uniformly usable for any graph algorithms."

![image](https://user-images.githubusercontent.com/1402801/77237986-f3712380-6bcc-11ea-8130-f7da8de3f4b3.png)


> Also do I assume that for algorithms where disconnected graphs don't make too much sense (like bfs) the benchmark graph datasets will be connected?

No, that assumption will not hold - you're expected to return "infinity" for the unreachable vertices.
E.g. in this traversal on the directed example, the BFS start from vertex 1, and vertices 2, 6, 7, 9 are unreachable:

![image](https://user-images.githubusercontent.com/1402801/77237972-d4729180-6bcc-11ea-87d4-e0100076770e.png)

**Comment by achalagarwal** [comment]
> If possible, you should put the preprocessing time into load column.

Yes, this is possible but then I would be overriding the load time set by graphalytics' v1.2.0 [here](https://github.com/ldbc/ldbc_graphalytics/blob/094f5a54039ea47cefb3316824f0ea13e1ceecc8/graphalytics-core/src/main/java/science/atlarge/graphalytics/execution/BenchmarkExecutor.java#L135)

> No, that assumption will not hold - you're expected to return "infinity" for the unreachable vertices.

Okay, got it, so the source vertex will always be specified in the benchmarks. 

Thanks!

**Comment by szarnyasg** [comment]
> Yes, this is possible but then I would be overriding the load time set by graphalytics' v1.2.0 here

Sure, that's not a problem as long as it has clear semantics and is reproducible by others.

**Comment by achalagarwal** [comment]
@szarnyasg 

Hi, do we still have the following restriction on CDLP? 

Excerpt from https://github.com/ldbc/ldbc_graphalytics/issues/167
> We would ideally leave all implementation details open so platforms can use whatever algorithm works best for them ... Exception to this rule is CDLP. To make validation of results possible, we use a deterministic definition of CDLP. Because this algorithm does not converge for all inputs, we opted to require a fixed number of iterations to be run. As a result, the reference (and required) output is intrinsically linked to our specific algorithm. As future work we may consider alternative definitions of what constitutes a "correct" implementation of CDLP (or even other CD algorithms), but this will not be included in Graphalytics v1.0.

Then along with WCC even CDLP would be on the pending list but I will personally try making sure that we get CDLP in as LPA is somewhat related to my main project!

**Comment by szarnyasg** [comment]
@achalagarwal yes, this restriction still holds in the current version (v1.0) of the benchmark. So there's a fixed number of iterations and in each iteration, vertices will be assigned the most frequent label propagated to them, with `min` used for breaking ties.

**Comment by achalagarwal** [comment]
@szarnyasg Could you please clarify a couple more things regarding label propagation as I believe that there are a lot of variables in this algorithm that can change the result.

1. I think graphalytics expects the algorithm described in https://arxiv.org/pdf/0709.2938.pdf (along with the min requirement), so instead of labelling a subset of the vertices initially, each vertex has a unique label initially? 

2. The order in which the vertices are updates does seem to affect the results too and is there a strict order graphalytics requires to enforce determinism there? 

Also, is there a wiki where could I find such details and more if I ever stumble across these issues in the future?

**Comment by szarnyasg** [comment]
@achalagarwal 

1. Initially, each vertex gets a unique label - its own identifier. See Section A.5 with the pseudocode https://ldbc.github.io/ldbc_graphalytics_docs/graphalytics_spec.pdf#page=39

2. In each iteration, all labels should be updated before going on to the next iteration. So if existing labels are L_{i}, then new labels L_{i+1} are computed based on L_{i} values. Once these are determined, we compute L_{i+2} based on L_{i+1} and so on.

3. There's a wiki but it's not heavily used: https://github.com/ldbc/ldbc_graphalytics/wiki. I agree that it'd be a good idea to open an FAQ here.

**Comment by achalagarwal** [comment]
Mentioning our LCC issue here as it is required in Graphalytics:
https://github.com/gunrock/gunrock/issues/496

**Comment by achalagarwal** [comment]
LCC update:

~~Louvain~~ Local Clustering Coefficient pull request created
